programs/bpf_loader/src/syscalls.rs

```diff
@@ -1329,6 +1343,79 @@ impl<'a> SyscallObject<BpfError> for SyscallMemset<'a> {
     }
 }
 
+// Ecrecover
+pub struct SyscallEcrecover<'a> {
+    base_cost: u64,
+    compute_meter: Rc<RefCell<dyn ComputeMeter>>,
+    loader_id: &'a Pubkey,
+}
+impl<'a> SyscallObject<BpfError> for SyscallEcrecover<'a> {
+    fn call(
+        &mut self,
+        hash_addr: u64,
+        recovery_id_val: u64,
+        signature_addr: u64,
+        result_addr: u64,
+        _arg5: u64,
+        memory_mapping: &MemoryMapping,
+        result: &mut Result<u64, EbpfError<BpfError>>,
+    ) {
+        question_mark!(self.compute_meter.consume(self.base_cost), result);
+
+        let hash = question_mark!(
+            translate_slice::<u8>(
+                memory_mapping,
+                hash_addr,
+                keccak::HASH_BYTES as u64,
+                self.loader_id,
+                true,
+            ),
+            result
+        );
+        let signature = question_mark!(
+            translate_slice::<u8>(memory_mapping, signature_addr, 64u64, self.loader_id, true,),
+            result
+        );
+        let ecrecover_result = question_mark!(
+            translate_slice_mut::<u8>(memory_mapping, result_addr, 64u64, self.loader_id, true,),
+            result
+        );
+
+        let message = match libsecp256k1::Message::parse_slice(hash) {
+            Ok(msg) => msg,
+            Err(_) => {
+                *result = Ok(1);
```

Comment:
I'm assuming that "1" here translates into something like "Parse error".  Can we define error types in program space for these for both Rust and C?